### ISSUE TITLE
Problems with Zip Files after upgrade to 7.0.0

### ISSUE THREAD

**Issue by maddox-DX** [owncloud/core#10001]
### Steps to reproduce
1. Login und go to files
2. select a couple of files download the generated zip file
3. try to decompress with Total Commander

### Expected behaviour
Files should be extracted

### Actual behaviour
Total Commander gives CRC warnings

### Server configuration
**Operating system:**
debian 7

**Web server:**
apache 2.2
**Database:**
postgres 9.1
**PHP version:**
5.4
**ownCloud version:** (see ownCloud admin page)
7.0.0
**Updated from an older ownCloud or fresh install:**
From 6.0.4

I got an hint from the Total Commander forum: http://ghisler.ch/board/viewtopic.php?t=40766 (in german)

The linked forum post says that the generated zip has two issues:

- the content is not compressed
- the file entries in the zip file don't provide valid crc data (and thats why total commander complains).

**Comment by DeepDiver1975** [comment]
Upstream project is still working on this issue -> OC8.1

**Comment by McNetic** [comment]
Update: There is now a deflate branch in PHPZipStreamer (https://github.com/McNetic/PHPZipStreamer/tree/deflate) that implements deflate compression when pecl_http (<2.0) is available. It could now be tested if MacOS is happy with Zip files created this way. I'm still investigating how to make it work without pecl_http.

**Comment by DeepDiver1975** [comment]
@LukasReschke @georgehrke my dear mac users - can you please test this branch? THX

**Comment by McNetic** [comment]
Another update: Looking at the deflate rfc, I noticed deflate can fall back to uncompressed blocks, which is totally simple to implement and imposes nearly no overhead compared to not using deflate at all, so I implemented an option to select deflate compression (COMPR::DEFLATE), but set compression level to none (COMPR::NONE). This should produce valid zip files (zip -t does not complain) with compression formally enabled, but without the need for pecl_http.

**Comment by DeepDiver1975** [comment]
awesome @McNetic THX

**Comment by LukasReschke** [comment]
Will test later.

**Comment by LukasReschke** [comment]
As I'm a big fan of digging in binary formats (not) I had the past few hours a look at the implementation and the ZIP specification (and realised that with > 10.8 even the store method might work and used `compress => COMPR::STORE` and `level => COMPR::NORMAL`)  but there seem to be some problems left:

- Extracting files not within a subdirectory fails because the general purpose purpose ADD flag (`0x0008`) seems to cause problems with OS X here. When setting the NONE  flag ( `0x0000`) those files can be extracted, so this case finally works.
- Extracting directories does not work at all and causes the same error when using DEFLATE and when not as well. Interestingly, a simple `zip` command (`zip --copy in.zip --out out.zip`) is able to repair the ZIP file. But for me it just looks like they added the CRC?

![screen shot 2015-02-13 at 19 39 58](https://cloud.githubusercontent.com/assets/878997/6193068/1f97d2e2-b3b8-11e4-8c0f-a802e556534f.png)

Files are available at: https://s3.owncloud.com/owncloud/public.php?service=files&t=c8d0394f050872efc8f32bd15d0a4d84

What is interesting is that http://pablotron.org/software/zipstream-php/ works just fine (after changing the version to `0x000A` because of some Finder madness). Can we maybe take a look at their implementation and see what makes the difference? (they seem to use `gzdeflate` and no ZIP64 but that is irrelevant for now)

@McNetic Do you happen to see on a first glance what coul dmake the difference there?

**Comment by LukasReschke** [comment]
I also compared some 7ZIP and ZipStreamer file using STORE, the output looks pretty much the same except the CRC (excerpt - more was not really relevant):

|                                                         | ZipStreamer                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | Working directory ZIP created with 7zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|---------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Local file header signature                             | 50 4B 03 04                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 50 4B 03 04                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Version needed to extract                               | 14 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 14 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| General purpose bit flag                                | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Compression method                                      | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification time                             | BA 74                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | BA 74                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification date                             | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| CRC-32                                                  | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Compressed size                                         | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Uncompressed size                                       | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| File name length (n)                                    | 05 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 05 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Extra field length (m)                                  | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File name                                               | 74 65 73 74 2F                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | 74 65 73 74 2F                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| Local file header signature                             | 50 4B 03 04                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 50 4B 03 04                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Version needed to extract (minimum)                     | 0A 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 0A 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| General purpose bit flag                                | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Compression method                                      | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification time                             | 9C 5B                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 9C 5B                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification date                             | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| **CRC-32**                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 5E AC 80 ED                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| **Compressed size**                                     | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | A3 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| **Uncompressed size**                                   | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | A3 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| File name length (n)                                    | 10 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 10 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Extra field length (m) (maybe also filename - not sure) | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File name                                               | 74 65 73 74 2F 77 65 6C 63 6F 6D 65 2E 74 78 74                                                                                                                                                                                                                                                                                                                                                                                                                                                          | 74 65 73 74 2F 77 65 6C 63 6F 6D 65 2E 74 78 74                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| Data                                                    | 57 65 6C 63 6F 6D 65 20 74 6F 20 79 6F 75 72 20 6F 77 6E 43 6C 6F 75 64 20 61 63 63 6F 75 6E 74 21 0A 0A 54 68 69 73 20 69 73 20 6A 75 73 74 20 61 6E 20 65 78 61 6D 70 6C 65 20 66 69 6C 65 20 66 6F 72 20 64 65 76 65 6C 6F 70 65 72 73 20 61 6E 64 20 67 69 74 20 75 73 65 72 73 2E 20 0A 54 68 65 20 70 61 63 6B 61 67 65 64 20 61 6E 64 20 72 65 6C 65 61 73 65 64 20 76 65 72 73 69 6F 6E 73 20 77 69 6C 6C 20 63 6F 6D 65 20 77 69 74 68 20 62 65 74 74 65 72 20 65 78 61 6D 70 6C 65 73 2E 0A 0A | 57 65 6C 63 6F 6D 65 20 74 6F 20 79 6F 75 72 20 6F 77 6E 43 6C 6F 75 64 20 61 63 63 6F 75 6E 74 21 0A 0A 54 68 69 73 20 69 73 20 6A 75 73 74 20 61 6E 20 65 78 61 6D 70 6C 65 20 66 69 6C 65 20 66 6F 72 20 64 65 76 65 6C 6F 70 65 72 73 20 61 6E 64 20 67 69 74 20 75 73 65 72 73 2E 20 0A 54 68 65 20 70 61 63 6B 61 67 65 64 20 61 6E 64 20 72 65 6C 65 61 73 65 64 20 76 65 72 73 69 6F 6E 73 20 77 69 6C 6C 20 63 6F 6D 65 20 77 69 74 68 20 62 65 74 74 65 72 20 65 78 61 6D 70 6C 65 73 2E 0A 0A |
| **???**                                                 | 5E AC 80 ED A3 00 00 00 A3 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| Central directory file header signature                 | 50 4B 01 02                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 50 4B 01 02                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Version made by                                     |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | 3F 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Version needed to extract (minimum)                     | 14 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 14 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| General purpose bit flag                                | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Compression method                                      | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification time                             | BA 74                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | BA 74                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File last modification date                             | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 4D 46                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| CRC-32                                                  | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Compressed size                                         | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| Uncompressed size                                       | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 00 00 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| File name length (n)                                    | 05 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 05 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Extra field length (m)                                  | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 24 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| File comment length (k)                                 | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Disk number where file starts                           | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| Internal file attributes                                | 00 00                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | 00 00

**Comment by McNetic** [comment]
First of all, thank you for your effort.
 * For compress == COMPR::STORE, level has no meaning. And we only had COMPR::STORE up until now, so I thought the idea was to use COMPR::DEFLATE (with level = COMPR::NONE if you don't have pecl_http available) to see if MacOS would like that.
 * I'm not sure I understand the part about files not in subdirectories. If I remember correctly, owncloud will always put all files in the zip into one top-level directory? Or did you test manually, without owncloud? The problem about the ADD flag is, that I have to provide the CRC and compressed size in the file header, so I would have to make two passes over the files, or hold them in memory. I also wonder that unsetting the ADD flag works at all. If I do this, unzip -t complains about "invalid compressed data to inflate". Maybe I misunderstood what you did?
 * I had a look at that pablotron implementation (not sure I found this one when looking for a solution to the problem about two years ago), and for a moment, I wondered about my sanity when seeing the use of gzdeflate(). I'm still puzzled how they could not have noticed the problem with gzdeflate(), as they even have a special handling of "large" files. I suspect you only tried small files with their implementation. The problem arises, when the file is split in chunks to process, so the file must be larger than the chunk size (~1MB for ZipStreamer, configurable but default 20MB for pablotrons ZipStream) to trigger the problem. Every output of gzdeflate() contains a "final" block defined in the deflate specification, which makes concatenating the output impossible. unzip -t complains about bad CRC for files generated in that way (I checked with pablotrons ZipStream to confirm still it really does ;-)).
Aside from that, they in fact do two passes over all files (and hold "small" files in memory), so they can add the CRC/compressed sizes to the header.

So, what I'm not sure you did, but I'd like to know: do zip files work with one top-level directory everything is in, and compress == COMPR::DEFLATE and variying settings of level?

**Comment by LukasReschke** [comment]
> First of all, thank you for your effort.

Nah, nothing to thank. You are the one doing all the hard work here. I'm just doing some monkey testing :smiley: 

> * For compress == COMPR::STORE, level has no meaning. And we only had COMPR::STORE up until now, so I thought the idea was to use COMPR::DEFLATE (with level = COMPR::NONE if you don't have pecl_http available) to see if MacOS would like that.

Yep. Unfortunately it doesn't always like that it seems. Apple is doing great things…

>* I'm not sure I understand the part about files not in subdirectories. If I remember correctly, owncloud will always put all files in the zip into one top-level directory? Or did you test manually, without owncloud? The problem about the ADD flag is, that I have to provide the CRC and compressed size in the file header, so I would have to make two passes over the files, or hold them in memory. I also wonder that unsetting the ADD flag works at all. If I do this, unzip -t complains about "invalid compressed data to inflate". Maybe I misunderstood what you did?

There is a way to get files not in a top-level directory. To do that create two text files in the top directory and select them both with the checkbox. In the upper right of the table there will be a "Download" button then. In that case without the `ADD` flag it worked fine. However, that seems more like Finder accepting some probably not correct values but failing with correct ones :see_no_evil: 

>  * I had a look at that pablotron implementation (not sure I found this one when looking for a solution to the problem about two years ago), and for a moment, I wondered about my sanity when seeing the use of gzdeflate(). I'm still puzzled how they could not have noticed the problem with gzdeflate(), as they even have a special handling of "large" files. I suspect you only tried small files with their implementation. The problem arises, when the file is split in chunks to process, so the file must be larger than the chunk size (~1MB for ZipStreamer, configurable but default 20MB for pablotrons ZipStream) to trigger the problem. Every output of gzdeflate() contains a "final" block defined in the deflate specification, which makes concatenating the output impossible. unzip -t complains about bad CRC for files generated in that way (I checked with pablotrons ZipStream to confirm still it really does ;-)).
Aside from that, they in fact do two passes over all files (and hold "small" files in memory), so they can add the CRC/compressed sizes to the header.

:insert_sighing_meme_about_all_this_madness_caused_by_apple_doing_some_things_terribly_wrong_since_ages:

> So, what I'm not sure you did, but I'd like to know: do zip files work with one top-level directory everything is in, and compress == COMPR::DEFLATE and variying settings of level?

Let me post some test combinations here using `COMPR::DEFLATE` and ZIP64 support disabled:
- zipping folder `/test/` which contains no files below
  - `COMPR::NONE` :white_check_mark: 
  - `COMPR::NORMAL` :white_check_mark: 
  - `COMPR::MAXIMUM` :white_check_mark: 
  -  `COMPR::SUPERFAST` :white_check_mark: 
- zipping single files `/welcome.txt` and `/welcome1.txt`
  - `COMPR::NONE` :x: 
  - `COMPR::NORMAL` :x: 
  - `COMPR::MAXIMUM` :x: 
  -  `COMPR::SUPERFAST` :x: 
- zipping folder `/test/` which contains files below
  - `COMPR::NONE` :x: 
  - `COMPR::NORMAL` :x: 
  - `COMPR::MAXIMUM` :x: 
  -  `COMPR::SUPERFAST` :x: 

The confusing thing is the error thrown in the console `Error -60005 creating authorization` which on OS X usually indicates that a process has not the correct permissions to create something which is why I first suspected that this has to do with some path values, however when looking at the working and the not working ZIP (as compared above) this makes no sense at all… :disappointed:

**Comment by LukasReschke** [comment]
@McNetic Btw. I just got access from @mmayer to a OS X VM (accessible via RDP etc…) which we can use for unit testing… I guess he won't have anything against if I grant you access to it so that you could do some live testing and ZIP support might work again?

If you're interested you know my mail address (or can look it up on my profile) ;-)

**Comment by iliyasbasir** [comment]
One of our enterprise customer having this issue, can you advise me do you have fix/Patch for this issue ? If “yes”...Can we have our internal QA test fix; we can provide Mac access if he/she needed. 

#### Here is detail from case
“We are having a new issue since migrating to OC EE 7.0.5 - Specifically on Mac OS clients, people are reporting when multiple files or directories are downloaded from ownCloud, it results in a zip file that can not be opened on Mac OS system. 

I'm wondering if there is a workaround we can implement server side until we can get to OC EE 8.2, and/or what your official recommendation is? 

We have a lot of users which have Macs, and multi-file downloads are effectively broken for them right now.”

Thank you for help

**Comment by LukasReschke** [comment]
Ilyias, we do not have a patch at the moment that works on OS X reliably. For the moment users can use another unzipper such as Keka or StuffIt.

**Comment by LukasReschke** [comment]
(as also stated in the discussion above)

**Comment by MorrisJobke** [comment]
@iliyasbasir There is nothing more I can add.

**Comment by McNetic** [comment]
Ok, I found some time to have a look at the OSX problem with the VM provided by @mmayer, myself. I also found other sources complaining about the OSX incompatibility with some zip features. So my conclusion is that the OSX finder cannot handle zip general purpose flag bit 3 set (aka streaming zip files, part of official zip standard since 1993). I can not handle this problem in ZipStreamer itself, as it does also work on networked streams, which don't support fseek(). Only potential upside: maybe Finder does in fact support zip64 - I haven't tried yet.

Now, to solve the problem for owncloud, I could add an option to provide file size and crc32 to ZipStreamer from owncloud directly. There are still some questions how this whole thing should be designed:
* should owncloud try to detect osx and fall back to the proposed solution while retaining the current behaviour otherwise?
* could owncloud calculate and cache the crc32 (and size) for each file on upload (when missing, it wil still be needd to calculate on first time it is downloaded via zip file)?
* for crc32 calculation on download: is (and will be) the stream provided by Filesystem class always seek()able?

**Comment by McNetic** [comment]
I just noticed: Technically, this issue was about a small bug that was fixed in Zipstreamer months ago. The OSX issues are not even mentioned in the issue report. There is also at least one separate issue about the OSX problems (#13669). I'm not quite sure what to do about that.
One *could* implement the current Zipstreamer version (I'd do a release if necessary, it should not be a step backward related to the OSX issues), and close this issue, and follow up in #13669.

**Comment by LukasReschke** [comment]
Please apologize, @McNetic, that I get back to you that late. This simply slipped through on my huge backlog :-/

> Now, to solve the problem for owncloud, I could add an option to provide file size and crc32 to ZipStreamer from owncloud directly. There are still some questions how this whole thing should be designed:
> should owncloud try to detect osx and fall back to the proposed solution while retaining the current behaviour otherwise?

Personally, I'd like to avoid a fallback as much as possible. This leads to confusing behaviour as soon as people start mailing / sharing the generated ZIP files. Unlikely, but can happen. – Yet, I assume that this would require quite more resources and we should do this though. Requires some measuring on my end.

> could owncloud calculate and cache the crc32 (and size) for each file on upload (when missing, it wil still be needd to calculate on first time it is downloaded via zip file)?

I think we should avoid this. Files can also be modified with other means than through ownCloud (i.e. a mounted Windows Network Drive) and going this route is probably opening a can of worms as we need to ensure that all backends behave properly. :boom:

(and there are actually some cases of custom backends :-()

> for crc32 calculation on download: is (and will be) the stream provided by Filesystem class always seek()able?

I'm not completely sure about that. @icewind1991 @DeepDiver1975 can one of you folks please comment on this? THX. – Also input on the other questions is appreciated.

> I just noticed: Technically, this issue was about a small bug that was fixed in Zipstreamer months ago. The OSX issues are not even mentioned in the issue report. There is also at least one separate issue about the OSX problems (#13669). I'm not quite sure what to do about that.
> One could implement the current Zipstreamer version (I'd do a release if necessary, it should not be a step backward related to the OSX issues), and close this issue, and follow up in #13669.

We can do that if that would help. – Tagging a new release would also be appreciated. (we don't ship master when we can avoid it)

**Comment by cdamken** [comment]
After upgrade to 7.0.5 does not work either.

@MorrisJobke 

00002808

**Comment by cdamken** [comment]
@maddox-DX @LukasReschke Any news?

**Comment by LukasReschke** [comment]
According to the issue tracker: ![2015-06-25_10-44-20](https://cloud.githubusercontent.com/assets/878997/8350600/4c2aef70-1b27-11e5-9b20-b8734d22e5a8.png)

I have asked questions in https://github.com/owncloud/core/issues/10001#issuecomment-102022858 which are unanswered and unless those are answered we can't proceed.

**Comment by DeepDiver1975** [comment]
@cdamken this issue is scheduled for 8.2 - there can be no progress ...

![bildschirmfoto von 2015-06-25 12-08-49](https://cloud.githubusercontent.com/assets/1005065/8352119/0fff29e2-1b33-11e5-8561-85d56b367345.png)

**Comment by cdamken** [comment]
@DeepDiver1975  Ok, thanks

@MorrisJobke 
00002808
00003426

**Comment by MTRichards** [comment]
After all this, can we fix this for 8.2? Do we know what we want to do?

**Comment by MTRichards** [comment]
@cmonteroluque while this is tagged, it is in?

**Comment by DeepDiver1975** [comment]
> After all this, can we fix this for 8.2? Do we know what we want to do?

we rely on an upstream project which is fighting this issue for a long time. The problem at the end is:
as soon as we fix it for one platform - it will fall a part on another.
Long story short: zip seems to be a nightmare with respect to it's specification.

If I remember correctly this issue was meanwhile reported to Apple as well - so maybe they fix it? :speak_no_evil:

**Comment by LukasReschke** [comment]
Well @McNetic has asked for questions which still are unanswered at https://github.com/owncloud/core/issues/10001#issuecomment-94480285 :speak_no_evil: 

> If I remember correctly this issue was meanwhile reported to Apple as well - so maybe they fix it? :speak_no_evil:

Yeah. Maybe in unicorn land :speak_no_evil:

**Comment by bboule** [comment]
While it does not solve all of the problems can we at least address the one use case (multiple file downloads) with a small change in behavior i.e. Serial downloads on multi file select rather than zip them up?  There are still a number of other impacted use cases (like downloading a folder) but perhaps we could provide some relief on one front?

**Comment by PVince81** [comment]
@bboule serial download is not possible because browsers only support downloading one file at a time. This is the reason why zip is used. Or are you aware of any new tech that could allow that ?

**Comment by DeepDiver1975** [comment]
> Well @McNetic has asked for questions which still are unanswered at #10001 (comment) :speak_no_evil:

I see no reliable way to store the crc checksum for every file - this is too much of a change.
Hmmm - this is really a pain ....

**Comment by scolebrook** [comment]
@bboule To my knowledge only FF supports multipart HTTP bodies allowing multiple downloads in one request. Perhaps there's a way for javascript to be used so the browser sends multiple separate requests. Could be security implications though with accidental (or deliberate) DOS.

So perhaps, offering both methods (make zip file and send it or stream) and allowing admins to choose via config.php which method is used under which circumstances (for selections of files, folders, selections over a certain size, etc) would be best. When a user base is Mac heavy, use the old way and just deal with needing the extra space to store the zips while they are downloaded. When there are few to no Mac users, take advantage of streaming.

If hell ever freezes over and Apple decides to invest 10 minutes of engineering effort into something that doesn't have flashy animations then we can leave the old method forever. But an end user doesn't care that it's Apple's fault. To them, it's broken and they want it fixed. Installing an alternate unarchive tool is just a band aid solution to them. Giving admins more power to handle the unique nature of each installation is better than simply waiting on Apple.

**Comment by DeepDiver1975** [comment]
we might want to have a look at : https://github.com/barracudanetworks/ArchiveStream-php

**Comment by PVince81** [comment]
@DeepDiver1975 "Only the Zip64 (version 4.5 of the Zip specification) format is supported.". Wasn't that the actual problem with Apple ?

**Comment by DeepDiver1975** [comment]
> Wasn't that the actual problem with Apple ?

yes - but there will be tar for mac then

**Comment by bboule** [comment]
@PVince81 Yes indeed but I was hoping we could control it via javascript magic (sounds like my hopes are dashed ;) )

**Comment by PVince81** [comment]
@bboule I read somewhere that you can trigger downloads by appending hidden iframes. Problem is that some browsers will pop up the "save as" dialog. Imagine this if the user selected 100 files...

**Comment by bboule** [comment]
I see your point  not ideal :)

**Comment by gig13** [comment]
@MTRichards @bboule 
I would like to request a backport to OC7 if possible.

**Comment by MorrisJobke** [comment]
> I would like to request a backport to OC7 if possible.

We don't even have a patch for master. Even an idea how to fix this properly is very hard :(

**Comment by guruz** [comment]
Why not generate different kind of zip files based on the client browser OS?
The usual use case is probably downloading a directory and this would enable that.

**Comment by MTRichards** [comment]
Can we do that? This would be ideal, and allow for future unknown OS support problems with certain file types.

**Comment by DeepDiver1975** [comment]
> Can we do that? This would be ideal, and allow for future unknown OS support problems with certain file types.

afaik not because MacOS does not support streamed zip archives - but we require zip streaming to have a low memory foodprint

**Comment by butonic** [comment]
I tried using the [deflate branch](https://github.com/McNetic/PHPZipStreamer/tree/deflate) and toyed with constructor options
```php
			$zip = new ZipStreamer([
				'zip64' => false,
				'compress'=>\ZipStreamer\COMPR::DEFLATE,
				'level'=>\ZipStreamer\COMPR::NORMAL
			]);

			$zip = new ZipStreamer([
				'zip64' => false,
				'compress'=>\ZipStreamer\COMPR::NONE
			]);

			$zip = new ZipStreamer([
				'zip64' => false,
				'compress'=>\ZipStreamer\COMPR::DEFLATE,
				'level'=>\ZipStreamer\COMPR::NONE
			]);
```
In all cases downloadad zips fail to open with the Archive Utility but allow me to use unzip (5.52 of 28 February 2005):
```
$ unzip download-11.zip
Archive:  download-11.zip
  inflating: Neue Textdatei.txt      
  inflating: üöäß⊂∫∀∃ℕℝ∂.txt  
  inflating: welcome.txt
$ zipinfo download-11.zip
Archive:  download-11.zip   1350 bytes   3 files
-rw-r--r--  4.5 unx     1487 bl defN  4-Aug-15 11:30 Neue Textdatei.txt
-rw-r--r--  4.5 unx        0 bl defN  4-Aug-15 11:30 üöäß⊂∫∀∃ℕℝ∂.txt
-rw-r--r--  4.5 unx      163 bl defN  4-Aug-15 11:30 welcome.txt
3 files, 1650 bytes uncompressed, 940 bytes compressed:  43.0%
$ zipinfo download-12.zip
Archive:  download-12.zip   2231 bytes   3 files
-rw-r--r--  4.5 unx     1658 bl stor  4-Aug-15 11:38 Neue Textdatei.txt
-rw-r--r--  4.5 unx        0 bl stor  4-Aug-15 11:38 üöäß⊂∫∀∃ℕℝ∂.txt
-rw-r--r--  4.5 unx      163 bl stor  4-Aug-15 11:38 welcome.txt
3 files, 1821 bytes uncompressed, 1821 bytes compressed:  0.0%
$ zipinfo download-13.zip
Archive:  download-13.zip   2256 bytes   3 files
-rw-r--r--  4.5 unx     1658 bl defN  4-Aug-15 11:41 Neue Textdatei.txt
-rw-r--r--  4.5 unx        0 bl defN  4-Aug-15 11:41 üöäß⊂∫∀∃ℕℝ∂.txt
-rw-r--r--  4.5 unx      163 bl defN  4-Aug-15 11:41 welcome.txt
3 files, 1821 bytes uncompressed, 1846 bytes compressed:  -1.4%
```

I was using the "Neue Textdatei.txt" for copy paste exchange ... thats why the size changed.

Setting `'zip64' => false` makes the zip file use 0x0a as the min version required, and I fail to understand what the remaining problem with the Archiver Utility is ... so ... I don't know ... progress?

**Comment by McNetic** [comment]
The native OSX Tools (unzip + archive utility) both don't support the zip64 extension, which is part of zip standard since at least 2006. As a bonus, the archive tool also does not support streamed zip files, which is part of zip standard since 1993. Zipstreamer does always use streamed zip - as the name suggests). I think the best way is to detect OSX clients, and disable large zip downloads (with a manual override stating that third party tools are required to unpack).

**Comment by butonic** [comment]
@DeepDiver1975 I guess detecting OSX and then disabling 64bit support in zipstreamer is the first step. If we can detect if the selected files require using 64bit we can threw en exception. That should solve this for the majority of cases.

**Comment by LukasReschke** [comment]
>  I guess detecting OSX and then disabling 64bit support in zipstreamer is the first step. If we can detect if the selected files require using 64bit we can threw en exception. That should solve this for the majority of cases.

It's not that easy. It's also about streamed ZIPs which are unsupported as well. We do need a completely different solution here thus.

**Comment by butonic** [comment]
OK, can we agree on OSX being broken beyond sanity? Good. Now, the only thing we *can* do for end users is providing a non zip64 download when the file size allows. That would allow them to use unzip. That's it. No more (archive utility still does not work), no less (makes downloading collections minimally usable).

That being said, IIRC someone mentioned using tar(.gz) for OSX? That approach would still require detecting osx ...

**Comment by LukasReschke** [comment]
.tar stuff: https://github.com/owncloud/core/issues/17930

That said: Yes, sounds feasible. - Maybe also a notification on the download page or so about this.

**Comment by MTRichards** [comment]
Excellent! Do we have a plan? Sounds like it...

**Comment by karlitschek** [comment]
Not sure what the plan is. Switching to tar makes things not easier. There are a ton of different incompatible tar dialects as I have discovered with our software download files. We use zip files now because it has better compatibility across systems.

**Comment by DeepDiver1975** [comment]
> Not sure what the plan is. Switching to tar makes things not easier. There are a ton of different incompatible tar dialects as I have discovered with our software download files.

We would need to test the created tar on Mac and see if it works on this platform.

Tar is no understood by Windows - but we dont care since tar will be a OSX only solution

**Comment by karlitschek** [comment]
ok. but we need testing of all the newer os x version. this is unfortunately often incompatible between tar versions. also tar has problems with different charsets, path lengths and so on.

**Comment by DeepDiver1975** [comment]
> also tar has problems with different charsets, path lengths and so on.

hmmm ... is there any archive format which works properly? unbelievable ...

**Comment by karlitschek** [comment]
Yes. I thought that something that old and stable as tar always works. unfortunately this is more fragile then I thought. but if this is only used on OS X then this can work with  proper testing.

**Comment by raimund-schluessler** [comment]
This sounds like the problems with ZIP unpacking in Total Commander which still does not work (oC 8.1.1 and TotalCommander 8.51a) won't be solved? Is this right?

**Comment by LukasReschke** [comment]
> This sounds like the problems with ZIP unpacking in Total Commander which still does not work (oC 8.1.1 and TotalCommander 8.51a) won't be solved? Is this right?

Can you please file this at https://github.com/McNetic/PHPZipStreamer/issues?

**Comment by cdamken** [comment]
> We use zip files now because it has better compatibility across systems.

How about using RAR? its second in popularity to .zip files and multi-platform.

**Comment by cdamken** [comment]
One of our users suggested to either make it an option to switch streaming zip files on or off, or use non-streaming ZIP files for files <4 GB

**Comment by DeepDiver1975** [comment]
> One of our users suggested to either make it an option to switch streaming zip files on or off, or use non-streaming ZIP files for files <4 GB

This is a nightmare when it comes down to server memory foot print and other server resources.

**Comment by DeepDiver1975** [comment]
> How about using RAR?

afaik not nativly supported on windows

**Comment by PVince81** [comment]
Please note that in 8.2 there will be an actions dropdown: this means it would be possible to allow multiple download formats in the menu without cluttering the UI too much:
![fileactions-bubble](https://cloud.githubusercontent.com/assets/277525/9107783/4d1dbb78-3c2a-11e5-87f4-2a58c7e8f479.png)

In that case we might want to have such dropdown in the public link page too when clicking on the button on the top right.

But this means users need to know what to choose.

That was just a side-note :smile:

**Comment by cdamken** [comment]
> afaik not nativly supported on windows

You are right, and the recommendation from Microsoft:
Windows does not natively support RAR files, but there are third-party programs that do. To extract files from a compressed RAR file, you need to use a program that is designed to work with RAR files.
Ref: http://windows.microsoft.com/en-us/windows/what-is-rar-file#1TC=windows-7

and zip is not native in Linux either:
Linux has both zip and unzip program. By default, these utilities are not installed. You can install zip/unzip tools from the shell prompt.
Ref:http://www.cyberciti.biz/tips/how-can-i-zipping-and-unzipping-files-under-linux.html

IMO, lets provide a solution with a well know format that can be open in all the devices that we support.

**Comment by butonic** [comment]
@cdamken Installing third party software defeats the purpose. most linux desktops come with a ui tool that handles zip. gnome file-roller certainly does. Windows zip support is slow as hell ... but works. OSX is the problem. Using your logic, they can just install [The Unarchiver](https://itunes.apple.com/de/app/the-unarchiver/id425424353?mt=12).

**Comment by cdamken** [comment]
I was only telling about rar because someone pointed tar too.

@butonic you are right, lets only concentrate in zip and mac in this issue.

**Comment by butonic** [comment]
@cdamken The idea is to detect OSX and only then use tar.

**Comment by jnweiger** [comment]
In https://github.com/McNetic/PHPZipStreamer/issues/16#issuecomment-94475010 @McNetic indicates a solution. Added calls to provide crc() and size(). If I understand correctly, this should make the ZipStreamer OSX compatible. But that issue was closed wontfix. :-(

**Comment by cdamken** [comment]
@butonic the customers wants to have zip and not tar files. I already suggested  "The Unarchiver" and many other workarounds.

**Comment by butonic** [comment]
@cdamken Are you sure he wants zip? Or does he just want to be able to open the downloaded archive with the native osx application. Those are two different things and tar would allow that.

**Comment by cdamken** [comment]
@butonic I will ask all of them, I'll write again with their expectations.

**Comment by MTRichards** [comment]
Since this also affects another customer now, we need to get to something that works for Mac. The key is that it work natively on Mac, simple click and download. Whether tar or zip doesn't matter, as long as it works on Mac natively. The specific archive format should be all but invisible to the end user.

Priorities:
1) get some form of compression working seamlessly on all Macs so the product works
2) maybe worry about which format this customer specifically wants

**Comment by DeepDiver1975** [comment]
So we go with tar and autodetection of the client platform.



@MTRichards would it be enough to have a solution in the scope of 8.2 or are you seeking for backport to 8.1 or even earlier versions?

background: the affected code base will change in the scope of 8.2 - backport would mean: implement it again.

**Comment by MTRichards** [comment]
Well, unfortunately, we have a lot of customers (including our very own instance) where this isn't working properly. Since there are a lot of open tickets, and people are staying on 6 because of this, and we are dropping 6 in 2 months, and 8.2 isn't out until November....I don't see an alternative to at least 1 backport. I am OK if we only backport to 8.0, but we need to do something to resolve the bug reports if we can.

**Comment by karlitschek** [comment]
Hey guys. I don't think we want to support different format for the user or admin to choose. the user only wants to have exactly one format that works. Maybe we use zip streamer for all platforms except mac where we use the old zip code or tar? But al transparently and automatic without any options please.

**Comment by MTRichards** [comment]
>But al transparently and automatic without any options please. 

YES!

**Comment by DeepDiver1975** [comment]
> But al transparently and automatic without any options please.
> 
> YES!

This was the idea from the beginning of the tar discussion:
Detect the client platform and if osx send them a tar.

Just in case this got lost: https://github.com/owncloud/core/issues/17930#issuecomment-125590996

**Comment by cdamken** [comment]
@butonic 
> @cdamken Are you sure he wants zip?

Confirmed: They want zip! all other solution is just a workaround.

**Comment by MorrisJobke** [comment]
> They want zip! all other solution is just a workaround.

And zip is broken on Mac OS X ... so for files bigger than 4GB this will never work except Apple fixes it :(

**Comment by MTRichards** [comment]
We can't have folks unclear on the details making blanket statements like "it is a workaround". Sure, it is a workaround for APPLE. End of Story.

We are doing .tar first so it works. End of story. @DeepDiver1975 already slated .tar, done.

**Comment by karlitschek** [comment]
yep. we do what @MTRichards and @DeepDiver1975 proposed. done

**Comment by McNetic** [comment]
@raimund-schluessler: The problems with TotalCommander were indeed an (unrelated) bug in PHPZipStreamer and are fixed in the current development tree, which is up to now not bundled with owncloud. So, it will work in TotalComander in the future.

**Comment by raimund-schluessler** [comment]
@McNetic Great, thank you!

**Comment by michaelstingl** [comment]
@MorrisJobke Will 7.0.9 also generate TAR?

**Comment by MorrisJobke** [comment]
> @MorrisJobke Will 7.0.9 also generate TAR?

Currently there isn't even a solution for master.

**Comment by DeepDiver1975** [comment]
Gentlemen hold your horses. I even dont know if this can make it into 8.2

**Comment by McNetic** [comment]
Just to make sure I am not longer an obstacle in getting fixes and improvements into this feature, I just did a release (v0.7) of ZipStreamer, which fixes all known issues except of course the bugs in OSX. I think this version could be bundled with Owncloud as soon as possible.
As already pointed out in the past, there is a small API change in ZipStreamer since the last release. It should be pretty straightforward to adopt. If you need any help or I forgot anything in doing my release, please tell me.

**Comment by cmonteroluque** [comment]
@DeepDiver1975 check

**Comment by DeepDiver1975** [comment]
9.0 no time left from my pov. In addition the tar stream code looks evil.

**Comment by gig13** [comment]
@bboule
SF 3561

**Comment by McNetic** [comment]
I created (hopefully correct) pull requests to import the changes in owncloud/core#19018 and owncloud/3rdparty#196. This also enables zipstreamer to be loaded and updated via composer.

**Comment by DeepDiver1975** [comment]
> I created (hopefully correct) pull requests to import the changes in owncloud/core#19018 and owncloud/3rdparty#196. This also enables zipstreamer to be loaded and updated via composer.

Awesome - THX a lot!

**Comment by MTRichards** [comment]
Oooh...does this fix zip on Mac?!? (waiting for a party here!)

**Comment by DeepDiver1975** [comment]
> Oooh...does this fix zip on Mac?!? (waiting for a party here!)

no :crying_cat_face:

**Comment by MTRichards** [comment]
![disappointed minion](https://cloud.githubusercontent.com/assets/1818977/9858094/af93f21e-5aec-11e5-8452-8472980f25cd.jpg)

**Comment by McNetic** [comment]
Can.. not... resist...: fixing zip on Mac has to be done by Apple. Luckily, you'll get an OS update within the next days/weeks. Unfortunately, that simple fix for a years old bug will probably not be included :-(.

Back to topic: As already pointed out earlier, this bug report is technically not about the OSX issues, so I suggest closing this one when the pull request is properly closed and the actual issue in the original bug report is thus fixed.
There is a seperate bug report (#13669) for the OSX issues, which I suggest should be used for the further OSX discussion.

**Comment by michaelstingl** [comment]
@McNetic Probably not fixed in Mac OS X El Capitan :crying_cat_face: :crying_cat_face: :


![cursor_and_os_x_10_11](https://cloud.githubusercontent.com/assets/214010/9859116/ed5560f0-5b24-11e5-8826-66e9bb7bd193.png)
(Version 10.11 Beta 8)

**Comment by karlitschek** [comment]
The problem is on Apples side. Definitely! For sure!
Now that we have this clarified let me also say that the mission of ownCloud is to provide awesome user experience for our users. So we have to build a workaround. Unfortunately if we care about our users. Which we do.

**Comment by DeepDiver1975** [comment]
> Which we do.

maybe I can come up with a PR to introduce the tar stream - needs some coding ... https://github.com/owncloud/core/issues/10001#issuecomment-139169447

**Comment by cmonteroluque** [comment]
@DeepDiver1975 Any update? This was just raised up as a priority for the Mac

**Comment by DeepDiver1975** [comment]
this is @VicDeo's current WIP PR to get this fixed #19318

**Comment by DeepDiver1975** [comment]
fixed by https://github.com/owncloud/core/pull/19318

**Comment by cdamken** [comment]
AWESOME!!! Thanks guys!! You made my day!!

**Comment by MTRichards** [comment]
Woohoo!
Thanks everyone.

**Comment by MTRichards** [comment]
Backport @karlitschek ?

**Comment by karlitschek** [comment]
Awesome!

**Comment by karlitschek** [comment]
@DeepDiver1975 We have to backport this at least to 8.1, better also 8.0

**Comment by DeepDiver1975** [comment]
https://github.com/owncloud/core/pull/19318#issuecomment-143876938
